pkg/kubelet/network/kubenet/kubenet.go

```diff
@@ -0,0 +1,215 @@
+	if err != nil {
+		return fmt.Errorf("Error building CNI config: %v", err)
+	}
+
+	res, err := plugin.cniConfig.AddNetwork(plugin.netConfig, rt)
+	if err != nil {
+		return fmt.Errorf("Error adding container to network: %v", err)
+	}
+	if res.IP4 == nil {
+		return fmt.Errorf("CNI plugin reported no IPv4 address for container %v.", id)
+	}
+
+	plugin.podCIDRs[id] = res.IP4.IP.String()
+
+	return nil
+}
+
+func (plugin *kubenetNetworkPlugin) TearDownPod(namespace string, name string, id kubecontainer.DockerID) error {
+	runtime, ok := plugin.host.GetRuntime().(*dockertools.DockerManager)
+	if !ok {
+		return fmt.Errorf("Kubenet execution called on non-docker runtime")
+	}
+	netnsPath, err := runtime.GetNetNs(id.ContainerID())
+	if err != nil {
+		return err
+	}
+
+	rt := buildCNIRuntimeConf(name, namespace, id.ContainerID(), netnsPath)
+	if err != nil {
+		return fmt.Errorf("Error building CNI config: %v", err)
+	}
+
+	if err := plugin.cniConfig.DelNetwork(plugin.netConfig, rt); err != nil {
+		return fmt.Errorf("Error removing container from network: %v", err)
+	}
+
+	// no cached CIDR is Ok during teardown
+	cidr, ok := plugin.podCIDRs[id]
+	delete(plugin.podCIDRs, id)
+
+	if ok {
+		glog.V(2).Infof("Removing pod CIDR %s from shaper", cidr)
+		if err := plugin.shaper.Reset(cidr); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// TODO: Use the addToNetwork function to obtain the IP of the Pod. That will assume idempotent ADD call to the plugin.
+// Also fix the runtime's call to Status function to be done only in the case that the IP is lost, no need to do periodic calls
+func (plugin *kubenetNetworkPlugin) Status(namespace string, name string, id kubecontainer.DockerID) (*network.PodNetworkStatus, error) {
+	cidr, ok := plugin.podCIDRs[id]
+	if !ok {
+		return nil, fmt.Errorf("No IP address found for pod %v", id)
+	}
+
+	ip, _, err := net.ParseCIDR(strings.Trim(cidr, "\n"))
+	if err != nil {
+		return nil, err
+	}
+	return &network.PodNetworkStatus{IP: ip}, nil
+}
+
+func buildCNIRuntimeConf(podName string, podNs string, podInfraContainerID kubecontainer.ContainerID, podNetnsPath string) *libcni.RuntimeConf {
+	glog.V(4).Infof("Kubenet: using netns path %v", podNetnsPath)
+	glog.V(4).Infof("Kubenet: using podns path %v", podNs)
+
+	return &libcni.RuntimeConf{
+		ContainerID: podInfraContainerID.ID,
+		NetNS:       podNetnsPath,
+		IfName:      DefaultInterfaceName,
+	}
+}
+
+func (plugin *kubenetNetworkPlugin) PodCIDRChanged(cidr string) {
+	var err error
+
+	if err := ensureIPTablesMasqRule(); err != nil {
+		err = fmt.Errorf("Error on adding ip table rules: %v", err)
+		glog.Error(err)
+	}
+
+	if len(cidr) == 0 {
+		err = fmt.Errorf("PodCIDR not set; will not configure %s right now", BridgeName)
+		glog.Warning(err)
+	} else if err := plugin.updateBridgeCidr(cidr); err != nil {
+		err = fmt.Errorf("Error configuring %s: %v", BridgeName, err)
+		glog.Error(err)
+	}
+
+	plugin.host.SetNetworkState(err)
```

Comment:
I think it would be better to return an error and let the host decide what to do with it, rather than expose this control to plugins directly.